### PR TITLE
Add `->*` operator for non-member-pointers

### PR DESCRIPTION
**Pull request by zasdfgbnm**
In the future, we will be able to do something like:
```C++
PolymorphicValue val = ...;

// access style 1
auto x = val->*&TensorMetaData::data;  // x is a void*
val->*&TensorMetaData::logical_size = ...;

// access style 2
auto y = val->*"data";  // y is a PolymorphicValue holding a Pointer
val->*"logical_size" = ...;
```

Where style 1 is static access and style 2 is dynamic access. For example, if you have a

```C++
struct S {
  int x;
  int y;
};
```
then `val->*&S::x` will give you an `const int&` for `x`, but `val->*"x"` will give you a `PolymorphicValue` holding the value of `x`.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build
!build

**Comment by zasdfgbnm** [comment]
!build

File: test/test_dynamic_type.cpp [hunk 1]
```diff
@@ -1178,26 +1179,26 @@ struct CD {
   }
 };
 
-using ABCD = DynamicType<NoContainers, A, B, CD>;
```
Comment (zasdfgbnm) [inline test/test_dynamic_type.cpp:1181]: Just moved into `DynamicTypeTest.MemberPointer` for better code organization.

**Comment by zasdfgbnm** [comment]
Err... Clang does not work:
```C++
/home/gaoxiang/Fuser3/test/test_dynamic_type.cpp:1248:18: error: right hand operand to ->* has non-pointer-to-member type 'const char[2]'
  static_assert(f->*"x" == 1);
                 ^  ~~~
```
Looks like this error message comes from https://github.com/llvm/llvm-project/blob/b8a9c50f22947284577a0a49038ea8ae5d45d6fa/clang/lib/Sema/SemaExprCXX.cpp#L5989-L5999

But my understanding for the C++ standard mismatch with clang. In that source code, clang referenced this: https://eel.is/c++draft/expr.compound#expr.mptr.oper

> The binary operator ->* binds its second operand, which shall be a prvalue of type “pointer to member of T” to its first operand, which shall be of type “pointer to U” where U is either T or a class of which T is an unambiguous and accessible base class[.](https://eel.is/c++draft/expr.compound#expr.mptr.oper-3.sentence-1) The expression E1->*E2 is converted into the equivalent form (*(E1)).*E2[.](https://eel.is/c++draft/expr.compound#expr.mptr.oper-3.sentence-2)

which says that the RHS of `->*` must be a member pointer. However, I don't think this rule applies to overloaded operators, so clang should not be checking that, because:

In the same chapter, the standard also says things like (in https://eel.is/c++draft/expr.shift)
> The operands shall be prvalues of integral or unscoped enumeration type and integral promotions are performed[.](https://eel.is/c++draft/expr.shift#1.sentence-2) The type of the result is that of the promoted left operand[.](https://eel.is/c++draft/expr.shift#1.sentence-3) The behavior is undefined if the right operand is negative, or greater than or equal to the width of the promoted left operand[.](https://eel.is/c++draft/expr.shift#1.sentence-4)

clearly, this chapter only applies to built-in operators, for example in `std::cout << std::string`, neither `std::cout` nor `std::string` is "prvalues of integral or unscoped enumeration type"

Indeed, at the beginning of this chapter, Note 2 (https://eel.is/c++draft/expr#pre-note-2), the standard clearly mentioned that:

> [Note [2](https://eel.is/c++draft/expr#pre-note-2): Operators can be overloaded, that is, given meaning when applied to expressions of [class](https://eel.is/c++draft/class) type or [enumeration](https://eel.is/c++draft/dcl.enum) type[.](https://eel.is/c++draft/expr#pre-2.sentence-1) Uses of overloaded operators are transformed into function calls as described in [[over.oper]](https://eel.is/c++draft/over.oper)[.](https://eel.is/c++draft/expr#pre-2.sentence-2) Overloaded operators obey the rules for syntax and evaluation order specified in [[expr.compound]](https://eel.is/c++draft/expr#compound), but the requirements of operand type and value category are replaced by the rules for function call[.](https://eel.is/c++draft/expr#pre-2.sentence-3) Relations between operators, such as ++a meaning a+=1, are not guaranteed for [overloaded operators](https://eel.is/c++draft/over.oper)[.](https://eel.is/c++draft/expr#pre-2.sentence-4) — end note]

So, in summary, I think my design for the interface is correct. I will move forward with this interface design, and find a workaround for clang.

**Comment by zasdfgbnm** [comment]
Indeed, I think I should just post a PR to remove this check from clang, so that in the future we do not have to workaround such issue again.

**Comment by zasdfgbnm** [comment]
clang build is fixed

**Review by naoyam — APPROVED** [review]
LGTM

**Comment by zasdfgbnm** [comment]
!build
